File: Java/56. Merge Intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        
        // use merge to keep track of the merged intervals
        int[] merge = intervals[0];
        List<int[]> result = new ArrayList<>();
        int[] cur = new int[2];
        
        for(int i=1; i<intervals.length; i++){
            cur = intervals[i];
            if(cur[0] <= merge[1]){
                merge[1] = Math.max(cur[1], merge[1]);
            }else{
                result.add(merge);
                merge = cur;
            }
        }
        
        //System.out.println(merge);
        //System.out.println(cur);
        result.add(merge);
        
        
        return result.toArray(new int[result.size()][]);
    }
}

File: Python3/56. Merge Intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key = lambda x: x[0])
        
        result = []
        i = 0
        merge = intervals[0]
        
        for i in range(1, len(intervals)):
            cur = intervals[i]
            # if the start of the current interval is smaller than
            # the previously merged interval, merge again
            # else, add merge to result and start a new merge from cur
            if cur[0] <= merge[1]:
                merge[1] = max(cur[1], merge[1])
            else:
                result.append(merge)
                merge = cur
        
        
        result.append(merge)
            
        return result
        